### PR TITLE
Create intro_sort.py

### PR DESCRIPTION
**Pull request by YeonJeongLee00**
### **Describe your change:**
added introsort algorithm implementation to sorts

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by mateuszz0000 — DISMISSED** [review]
Please add doctests and typehints

**Comment by YeonJeongLee00** [comment]
I added doctests and typehints. Can you check my code?

**Review by NumberPiOso — CHANGES_REQUESTED** [review]
Could you please add tests for different examples.
- Large numbers
- Small numbers
- Empty list
- Single list
- Floating-point numbers
- Strings

**Comment by YeonJeongLee00** [comment]
@NumberPiOso added more test

**Comment by YeonJeongLee00** [comment]
@mateuszz0000 Can you review my code?

**Comment by JiHyunSong** [comment]
@cclauss excuse me sir, can you reivew this code?

File: sorts/intro_sort.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
+"""
+Introspective Sort is hybrid sort (Quick Sort + Heap Sort + Insertion Sort)
+if the size of the list is under 16, use insertion sort
+https://en.wikipedia.org/wiki/Introsort
+"""
+import math
+
+
+def insertion_sort(array: list, start: int, end: int) -> list:
```
Comment (cclauss) [inline sorts/intro_sort.py:9]: This function has no tests.

File: sorts/intro_sort.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        temp_index = i
+        temp_index_value = array[i]
+        while temp_index != start and temp_index_value < array[temp_index - 1]:
+            array[temp_index] = array[temp_index - 1]
+            temp_index -= 1
+        array[temp_index] = temp_index_value
+    return array
+
+
+def heapify(array: list, index: int, heap_size: int) -> None:  # Max Heap
```
Comment (cclauss) [inline sorts/intro_sort.py:20]: This function has no tests.

File: sorts/intro_sort.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    if right_index < heap_size and array[largest] < array[right_index]:
+        largest = right_index
+
+    if largest != index:
+        array[index], array[largest] = array[largest], array[index]
+        heapify(array, largest, heap_size)
+
+
+def heap_sort(array: list) -> list:
```
Comment (cclauss) [inline sorts/intro_sort.py:36]: This function has no tests.

File: sorts/intro_sort.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        heapify(array, n, i)
+
+    for i in range(n - 1, 0, -1):
+        array[i], array[0] = array[0], array[i]
+        heapify(array, 0, i)
+
+    return array
+
+
+def median_of_3(
```
Comment (cclauss) [inline sorts/intro_sort.py:49]: This function has no tests.

File: sorts/intro_sort.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        return array[first_index]
+    elif (array[middle_index] > array[first_index]) != (
+        array[middle_index] > array[last_index]
+    ):
+        return array[middle_index]
+    else:
+        return array[last_index]
+
+
+def partition(array: list, low: int, high: int, pivot: int) -> int:
```
Comment (cclauss) [inline sorts/intro_sort.py:64]: This function has no tests.

File: sorts/intro_sort.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        j -= 1
+        while pivot < array[j]:
+            j -= 1
+        if i >= j:
+            return i
+        array[i], array[j] = array[j], array[i]
+        i += 1
+
+
+def sort(array: list):
```
Comment (cclauss) [inline sorts/intro_sort.py:79]: This function has no tests.

File: sorts/intro_sort.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    return insertion_sort(array, start, end)
+
+
+if __name__ == "__main__":
+    import doctest
+
+    doctest.testmod()
+
+    user_input = input("Enter numbers separated by a comma : ").strip()
+    unsorted = [int(item) for item in user_input.split(",")]
```
Comment (cclauss) [inline sorts/intro_sort.py:132]:
```suggestion
    unsorted = [float(item) for item in user_input.split(",")]
```

**Review by cclauss — CHANGES_REQUESTED** [review]
All functions should have tests.  Maybe create one or more global test_arrays and test all functions against those common datasets.

File: sorts/intro_sort.py (hunk 7 above)
Comment (YeonJeongLee00) [reply sorts/intro_sort.py]: @cclauss I modified this code and added tests to each function. 
I added only one case. Do I have to add more tests?

File: sorts/intro_sort.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        j -= 1
+        while pivot < array[j]:
+            j -= 1
+        if i >= j:
+            return i
+        array[i], array[j] = array[j], array[i]
+        i += 1
+
+
+def sort(array: list):
```
Comment (JiHyunSong) [inline sorts/intro_sort.py:108]: @YeonJeongLee00 return type hint 빠져있어요. bot feedback 도 꼭 확인하는 버릇을 들이시는게 좋습니다!

File: sorts/intro_sort.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    ['a', 'b', 'c', 'd', 'e']
+    """
+    if len(array) == 0:
+        return array
+    max_depth = 2 * math.ceil(math.log2(len(array)))
+    size_threshold = 16
+    return intro_sort(array, 0, len(array), size_threshold, max_depth)
+
+
+def intro_sort(array: list, start: int, end: int, size_threshold: int, max_depth: int):
```
Comment (JiHyunSong) [inline sorts/intro_sort.py:143]: type hint

File: sorts/intro_sort.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+
+def insertion_sort(array: list, start: int, end: int) -> list:
+    """
+    >>> array = [4, 2, 6, 8, 1, 7, 8, 22, 14, 56, 27, 79, 23, 45, 14, 12]
+
+    >>> insertion_sort(array, 0, len(array))
+    [1, 2, 4, 6, 7, 8, 8, 12, 14, 14, 22, 23, 27, 45, 56, 79]
+    """
+    for i in range(start, end):
```
Comment (cclauss) [inline sorts/intro_sort.py:16]: Let's also enable the caller to simply do `insertion_sort(my_array)` and get back the expected result.
```suggestion
def insertion_sort(array: list, start: int = 0, end: int = 0) -> list:
    """
    >>> array = [4, 2, 6, 8, 1, 7, 8, 22, 14, 56, 27, 79, 23, 45, 14, 12]

    >>> insertion_sort(array, 0, len(array))
    [1, 2, 4, 6, 7, 8, 8, 12, 14, 14, 22, 23, 27, 45, 56, 79]
    """
    end = end or len(array)
    for i in range(start, end):
```

**Review by cclauss — CHANGES_REQUESTED** [review]
This is looking really cool..  One last thing to try before we land it.  Thanks for your hard work!

File: sorts/intro_sort.py (hunk 10 above)
Comment (cclauss) [reply sorts/intro_sort.py:16]: What happens if start is a negative number or if end is larger than len(array) or start > end?

**Review by cclauss — APPROVED** [review]
Thanks much for your hard work here... I am going to accept and merge this but I still have reservations about `insertion_sort()`...

What happens if:
* start is a negative number
* if end is larger than len(array)
* start > end

If you wanted to address these issues in a future pull request then I would be happy to review.

**Pull request merged**